### PR TITLE
Delete project functionality!!!

### PR DESCRIPTION
**Pull request by sdevalapurkar**
# Overview

## This PR contains the following changes

- Delete project functionality!!! 🗑️
- Also wipe attachments associated to project and survey from s3
- Deal with observations and occurrences later (block observation table does get wiped, but detailed deletion of records under survey level will be dealt with later on)

## This PR contains the following types of changes

- [x] New feature (change which adds functionality)

## How Has This Been Tested?

- Locally

### CODE REVIEW

File: api/src/paths/project/{projectId}/delete.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+      if (!getSurveyIdsResult || !getSurveyIdsResult.rows) {
+        throw new HTTP400('Failed to get survey ids associated to project');
+      }
+
+      const surveyAttachmentS3Keys: string[] = Array.prototype.concat.apply(
+        [],
+        await Promise.all(
+          getSurveyIdsResult.rows.map((survey: any) => getSurveyAttachmentS3Keys(survey.id, connection))
+        )
+      );
```
Comment (NickPhura) [inline api/src/paths/project/{projectId}/delete.ts:101]: Can this not be simplified to something like:

```typescript
 const surveyAttachmentS3Keys: string[] = await Promise.all(
    getSurveyIdsResult.rows.map((survey: any) => getSurveyAttachmentS3Keys(survey.id, connection))
  );
```

Comment (sdevalapurkar) [reply api/src/paths/project/{projectId}/delete.ts:101]: it complains that they are incompatible types:
```
Type 'any[][]' is not assignable to type 'string[]'.
Type 'any[]' is not assignable to type 'string'.
```

Comment (sdevalapurkar) [reply api/src/paths/project/{projectId}/delete.ts:101]: it returns nested array that I am just flattening

File: database/src/migrations/release.0.18/api_delete_project.sql [hunk 2]
```diff
@@ -20,6 +20,7 @@ begin
   delete from survey_proprietor where s_id in (select id from survey where p_id = __project_id);
   delete from survey_attachment where s_id in (select id from survey where p_id = __project_id);
   delete from study_species where s_id in (select id from survey where p_id = __project_id);
+  delete from block_observation where s_id in (select id from survey where p_id = __project_id);
```
Comment (anissa-agahchen) [inline database/src/migrations/release.0.18/api_delete_project.sql:23]: nice!

**Review by anissa-agahchen — APPROVED** [review]
Looks good

**Pull request merged**